### ISSUE TITLE
Mint 20 BIOS/legacy installs create spurious FAT32 /boot/efi

### ISSUE THREAD

**Issue by SimonIremonger** [linuxmint/linuxmint#312]
When installing Mint-20 on a Legacy/BIOS system, and using the 'automatic' partition creation methods,
E.g. "Erase entire disk and install Linux Mint"
-or- "Install alongside existing OS"  [shrink some partition, etc].
In all cases, the mint 20 installer creates a FAT32 partition [MBR type 'b'] and mounts on /boot/efi  !!

fdisk -l output and /etc/fstab entry below
/dev/sda1  *       2048   1050623   1048576   512M  b W95 FAT32
/dev/sda2       1052670 976771071 975718402 465.3G  5 Extended
/dev/sda5       1052672 976771071 975718400 465.3G 83 Linux
UUID=2CB8-AFB6  /boot/efi       vfat    umask=0077      0       1

KEY notes:-
1) This is a change/regression from mint 19.3/before installers.
2) This partition ends up being 'empty' and disused!.
3) This is NOT created as type 0xef [EFI system partition], just 'normal' FAT32 type.

I can understand there being some benefit to system being such that it can be 'converted' to UEFI booting (or, rather, always support both modes of booting) and as such having an 'actual' uefi-system-partition could (sometimes) be useful (but may break legacy-booting other existing OS....).
In short, I'm fairly sure this is just a bug in the mint20 installer, currently harmless-ish due to the partition type not being 0xef but certainly confusing (and wastes 512mb disk space...).

Responses?  Fix for mint 20.1?

**Comment by Andy2051** [comment]
More info here:
[https://forums.linuxmint.com/viewtopic.php?f=46&t=326270&p=1855497&hilit=mint+20+legacy#p1855497](url)

**Comment by SimonIremonger** [comment]
clem says on testing new iso's:
"The /boot/efi partition thing.. I think that comes from partman. Unless you consider 512MB as being a significant amount of space it's only a cosmetic issue."
Does this need reporting upstream somewhere?

**Comment by insomniacno1** [comment]
New install of LM 20.2 Uma has this problem too, in my case, my old linux install Xubuntu 16.04 is on sda1 and I left there in order to be able to copy config files over, the installer asked me if i wanted to install LM side by side with old Xubuntu and I said Yes, then it made this strange FAT32 partition. I tried this on another SSD with windows 7, and it did the same thing.

sda2 is a Fat32 partition made by LM installation 

sda3 is extended partition and on sda5 LM is installed. 

You guys need to come out with an update or a fix for this, there is no need for a FAT32 partition on a legacy bios system. I all the year I used linux, this is the first time i have seen this done by an installer.

My system: Lenovo Thinkpad X300, Core2Duo 1.2Ghz, old non-UEFI bios

**Comment by SimonIremonger** [comment]
I believe, this is a very fiddly-to-patch and low-priority matter from the Mint side of things.  I suspect same issue in Mint 20.3 (beta ISO images are now on mirrors), .
If the issue exists in Ubuntu 22.04 beta/testing images this needs attention as it will then creep into Mint 21, etc...  Indeed you could help testing mint 21 Beta images (and any LMDE5 images) for the issue, it may well now have been sorted upstream so it won't creep into next Ubuntu LTS and not creep into next Mint21 series at all, with any luck!.

**Comment by Andy2051** [comment]
This is unlikely to get fixed, ubuntu are planning to dump ubiquity in favour of a new installer.
[https://www.debugpoint.com/2021/02/new-ubuntu-desktop-installer/](url)
When it hits the streets and makes its way into mint... Your guess is as good as mine.

**Comment by mmortal03** [comment]
> Unless you consider 512MB as being a significant amount of space it's only a cosmetic issue.

Unfortunately, it's not just a cosmetic issue. It also breaks legacy boot scenarios which previously *just worked*, such as migrating the entire disk image to an external hard drive or thumb drive.

**Comment by LukasThyWalls** [comment]
Hello.

This is still in happing 21/_Vannesa_, i did a full installation in a machine with BIOS (no UEFI), and i see that /boot/efi partition, but everything was working and i thought it is how it works now and that's it.

However, i had an issue with the hard disk, and copied the partitions to another disk to backup everything. I copied the partitions and not cloned them because the size was different and i need to resize them, but it's something i did before al later restore the bootloader with _boot-repair_ or something like that.

The problem is that _boot-repair_ (included in the Linux Mint 21 Live ISO) can't fix the bootloader because it thinks that i need to disable the compatibility mode for legacy BIOS in the UEFI, but i don't have any UEFI, it is an old BIOS, so i can't disable that, because **is** that. Other options shows the same issue, and funnily, if i put that don't use the /boot/efi partition, says there is a GPT, and i need a small partition (the /boot/efi one) to use and continue.

I contacted _boot-repair_ maintainers to see if they can do something or not in this case.

For me i thought to copy the original bootsector from the original disk to the new disk, hoping it works, but it still will have the issue in the future if i try a tool like boot-repair.

I'm not an expert with the bootloader, but i fix similar issues in the past but this time this seems completely wrong because i really don't know how this bootloader is installed (i don't know what really is in the _/boot/efi_ partition if it's not the UEFI bootloader) and i don't know exactly how to put it again to work.

**Comment by mmortal03** [comment]
> For me i thought to copy the original bootsector from the original disk to the new disk, hoping it works, but it still will have the issue in the future if i try a tool like boot-repair.

I wonder if it would boot if, say, you used an MBR/Legacy Live USB to delete that little partition, and then manually installed grub to the Mint partition using something similar to the Ubuntu "Terminal Method" from the following page (scroll down to it)?:
https://www.howtogeek.com/114884/how-to-repair-grub2-when-ubuntu-wont-boot/

Btw, I think when I was experimenting with this whole legacy BIOS, extra FAT32 partition situation the other day, when installing from scratch, I was manually setting up one big partition, then installing to that using the Something Else menu, setting the mount point to / rather than /boot, and ignoring the subsequent error prompts. I don't know which or any of that might help you here.

One person in the following pertinent thread mentioned that they were able to rename the efi relevant files/folders, which might help you work around whatever boot-repair's logic is? I don't know for sure on that, but this directly links to their post in the thread: https://forums.linuxmint.com/viewtopic.php?p=2192107#p2192107

**Comment by LukasThyWalls** [comment]
Thanks @mmortal03 for the info. With that (and more searches and i did tests this week) i finally boot up the system again.

I need to point out also that i learnt that not only is the issue described here in the OP, also i did my own mistakes. Simply is the first time i deal with GPT disks and it not easy to find how to proceed and what differences have when you need to fix one (almost everything refers to the old msdos/mbr partition tables or simply use boot-repair).

The OP issue i had is the same, in a 600GB disk with Linux Mint 21 completed formatted by the installer, with that 512MB /boot/efi partition, and what i did is to copy the partitions to a 2TB disk (i needed to change sizes and keep some data, that's why i copied instead of cloning them) and try to make boot it again. I did that before with MBR/MSDOS partition tables, so i thought i would be easy to do it again...
The mistake i did was something i didn't realized: Although i formatted both disks in the same linux computer first (with months between them), the 600GB have a MBR/MSDOS partition table (the one what i installed Linux Mint 21 and it created the /boot/efi partition), and the 2TB have a GPT partition table, so copying partitions and thinks that is going to work like that was a mistake for my part.

What i did to fix it was this (with the help of the provided links and my own tests):
- First, this works with a disk with GPT partition table AND using BIOS booting (no UEFI).
- I formatted the old "/boot/efi" partition created by Linux Mint (512 MB at the beginning of the disk) as a partition with no format and flag it with _bios_grub_ flag. (This partition can be 1MB it seems, but for future compatibility reason i keep the 512 MB partition).
- Disable it from fstab, it doesn't need to be mounted. The other ones (root "/" or "swap") needs to have the correct UUIDs, of course.
- Now boot-repair doesn't take the system as an UEFI system, so it tries to fix it.
- However, at least in my case, when it booted, the system didn't boot entirely and failed, although it went into maintenance mode with root terminal. With that i could do a "grub-install /dev/sda", reboot, and it worked!

---

First, only to say that maybe in my case i did some mistakes and it's not the issue posted what made my problem, because it's was the pass from MBR partitions to GPT partitions, although it confused me more because at first it shouldn't be a /boot/EFI partition in a MBR disk first.

Looking here an there, i found this is not an issue with Linux Mint, because other distros like KUbuntu, also makes that /boot/efi FAT partition (In the computer i'm using right now it's like that, it uses a GPT disk and it has an UEFI using BIOS Legacy, i should fix that next time).

It should be reported somewhere, because the installer should be using by default GPT when using the entire disk for the installation and then:
- If it's BIOS or UEFI Bios Legacy, create a unformatted partition (>1MB, but can be 512MB for future compatibility) with bios_grub flag, and then the rest of partitions and install.
- If it's UEFI, create the EFI partition (512 MB, FAT32), the rest of partitions and install.
What i don't know, is why it created an MBR/MSDOS partition table by default in my case, and the it create that /boot/efi partition that supposedly is not needed at all, as the in a MBR/MSDOS disk it's not needed at all, it's needed in a GPT disk.

Anyway thanks for the info, and sorry for deviate from the original issue!

**Comment by Andy2051** [comment]
Simply to say that this is now over a year old and no change.
No one assigned, no label, no projects, no milestones, no branches or pull requests.

While I accept that at some point support for legacy boot will stop (I hope Clem will give a lot of advance notice of this), it is still around and while it is, it should be supported. In addition to the problems first reported with LM20, the LM21 installer will now give you a dire warning about the system not booting if you do a "something else" install without an EFI partition to get around the bug. 

Instead of getting fixed, it has got worse.

**Comment by mjohenneken** [comment]
With the ending Windows 10 support. I have some people(non-tech savvy) interested in switching to Linux. However this is the first issue they are running into before they can experience working with Linux Mint. Typical scenario: Install Linux on an old Laptop to try and test it - et voila - you run into this issue and they give up. 

Explaining via Video Chat to do what @LukasThyWalls did is gonna be quite difficult.
